### DISCUSSION TITLE
Support on_conflict feature of sqlite?

### DISCUSSION THREAD

**Discussion by appflowy** [diesel-rs/diesel#2826] [Ideas]
pg crate has the **upsert**  which supplies the on_conflict, Is there any plan to add these features to the SQLite crate?

**Comment by weiznich** [comment]
That is already implemented on the master branch. https://docs.diesel.rs/master/diesel/upsert/index.html

**Reply by appflowy** [reply]
thx for reply so soon. 😁

**Comment by EFanZh** [comment]
It seem that bulk upsert does not work for SQLite. See the following code:

```rust
diesel::insert_into(...)
    .values(value)
    .on_conflict(...)
    .do_update()
    .set(...)
    .execute(...);
```

It works if `value` is an object that derives `Insertable`. But if `value` is an array (`Vec` or slice) of those objects, I get the following error:

```text
the trait `SupportsDefaultKeyword` is not implemented for `Sqlite`
```

**Reply by weiznich** [reply]
This is expected behaviour and directly related to the fact that sqlite does not support the `DEFAULT` keyword for insert statements. This means we cannot do perform multirow inserts with possibly missing value and therefore we fallback to separate insert statements for each row in this case to allow users to fall back to default values in this case. You are welcome to submit a concrete proposal how to handle this in a different way. 

Or to word it differently: It's not related to upsert, but that's a general issue with the current way we handle inserts for sqlite.

**Reply by EFanZh** [reply]
The latest version of SQLite supports bulk inserts natively:

```sql
INSERT INTO foo (x, y, z)
    VALUES (1, 3, 7), (4, 5, 6), (3, 6, 9)
    ON CONFLICT DO UPDATE
    SET y = excluded.y;
```

Is it possible to utilize this feature for implementing bulk upserting?

**Reply by weiznich** [reply]
It's not about SQLite not supporting bulk inserts (that is supported since quite long), but about SQLite not supporting default value clauses:

That's not possible to write:
```sql
INSERT INTO foo (x, y, z)
    VALUES (1, 3, DEFAULT), (4, DEFAULT, 6), (DEFAULT, 6, 9);
```

The only possibility to support something like this is to do a separate insert statement per row and skip the column set to default. Again if you know a better way feel free to write a complete proposal how to implement this as new discussion thread.

**Reply by EFanZh** [reply]
I am not very familiar with the current implementation of diesel, and I am don’t know how it can be done. Can I just propose a feature request for bulk upsert?

**Reply by weiznich** [reply]
>  Can I just propose a feature request for bulk upsert?

What would that change? I mean we are well aware of the fact that bulk inserts on sqlite are the way there are. If we would know how to improve them considering the limitations given by the sql implementation in sqlite, the requirement not to do remove features without replacement and the limitations given by the type system we would actually have fixed that already. That means, just opening a feature request will change nothing there. If you want to see support for this consider doing some research on possible solutions instead of expecting other people to do that for you.

**Comment by Nihlus** [comment]
I've been playing with upserts for a bit now, and I found that there does not appear to be any support for an untargeted ON CONFLICT with the sqlite backend. Would this be a bug report or a continuation of this feature request (specifically, for something akin to `on_conflict_do_nothing` but `on_conflict_do_update` instead).

**Reply by weiznich** [reply]
This is an old feature discussion thread. If you have a new support question please open a new thread in the [discussion forum](https://github.com/diesel-rs/diesel/discussions/categories/q-a) that includes all relevant information (diesel version, sqlite version, example code, error message, etc) to reproduce your issue.
